state/watcher.go

```diff
@@ -3218,14 +3218,152 @@ func ensureSuffixFn(marker string) func(string) string {
+}
+
 // watchEnqueuedActionsFilteredBy starts and returns a StringsWatcher
 // that notifies on new Actions being enqueued on the ActionRecevers
 // being watched.
 func (st *State) watchEnqueuedActionsFilteredBy(receivers ...ActionReceiver) StringsWatcher {
-	return newCollectionWatcher(st, colWCfg{
-		col:    actionNotificationsC,
-		filter: makeIdFilter(st, actionMarker, receivers...),
-		idconv: actionNotificationIdToActionId,
+	return newActionNotificationWatcher(st, true, receivers...)
+}
+
+// actionNotificationWatcher is a StringsWatcher that watches for changes on the
+// action notification collection, but only triggers events once per action.
+type actionNotificationWatcher struct {
+	commonWatcher
+	source   chan watcher.Change
+	sink     chan []string
+	filter   func(interface{}) bool
+	fireOnce bool
+}
+
+// newActionNotificationWatcher starts and returns a new StringsWatcher configured
+// with the given collection and filter function
+func newActionNotificationWatcher(backend modelBackend, fireOnce bool, receivers ...ActionReceiver) StringsWatcher {
+	w := &actionNotificationWatcher{
+		commonWatcher: newCommonWatcher(backend),
+		source:        make(chan watcher.Change),
+		sink:          make(chan []string),
+		filter:        makeIdFilter(backend, actionMarker, receivers...),
+		fireOnce:      fireOnce,
+	}
+
+	w.tomb.Go(func() error {
+		defer close(w.sink)
+		defer close(w.source)
+		return w.loop()
+	})
+
+	return w
+}
+
+// Changes returns the event channel for this watcher
+func (w *actionNotificationWatcher) Changes() <-chan []string {
+	return w.sink
+}
+
+func (w *actionNotificationWatcher) loop() error {
+	var (
+		changes []string
+		in      = (<-chan watcher.Change)(w.source)
+		out     = (chan<- []string)(w.sink)
+	)
+
+	w.watcher.WatchCollectionWithFilter(actionNotificationsC, w.source, w.filter)
+	defer w.watcher.UnwatchCollection(actionNotificationsC, w.source)
+
+	changes, err := w.initial()
+	if err != nil {
+		return err
+	}
+
+	var active map[string]struct{}
+	if w.fireOnce {
+		active = make(map[string]struct{})
+		for _, v := range changes {
+			active[v] = struct{}{}
+		}
+	}
+
+	for {
+		select {
+		case <-w.tomb.Dying():
+			return tomb.ErrDying
+		case <-w.watcher.Dead():
+			return stateWatcherDeadError(w.watcher.Err())
+		case ch := <-in:
+			updates, ok := collect(ch, in, w.tomb.Dying())
+			if !ok {
+				return tomb.ErrDying
+			}
+			if w.fireOnce {
+				if err := w.filterAndMergeIds(&changes, updates, active); err != nil {
+					return err
+				}
+			} else {
+				if err := w.mergeIds(&changes, updates); err != nil {
+					return err
+				}
+			}
+			if len(changes) > 0 {
+				out = w.sink
+			}
+		case out <- changes:
+			changes = []string{}
+			out = nil
+		}
+	}
+}
+
+func (w *actionNotificationWatcher) initial() ([]string, error) {
+	var ids []string
+	var doc actionNotificationDoc
+	coll, closer := w.db.GetCollection(actionNotificationsC)
+	defer closer()
+	iter := coll.Find(nil).Iter()
```

Comment:
this could get really messy if there's a large number of unpruned completed actions.
can we filter on pending or running...?

Reply:
to preserve backwards compatibility, we need to fire for pending and running, at least once. If the uniter crashes it should get the running actions it had started so it can error them.
number of pending/running actions is unbounded, but I believe this being for a legacy facades only we can accept that risk.

A) We could fire all initial events, then subsequently only fire when the notify document is missing a change field, but that requires a read for each notification. This also changes the semantics of the fire once for each action. 

B) A way around this is two collections, one for legacy watchers (fires only on pending) and one the new facade. This seems just as heavy as the current solution.

I'll implement option A and change all the semantics of fireOnce to notifyNewPending. But this solution isn't as robust as the current map or option B, future changes to the notify collection will need to keep in mind legacy behaviour.

This would have been substantially easier if the mongo collection watchers send the new doc state in the Change value (or a way to read the document in the past with revno)

Reply:
Pending and running yes. I was suggesting that'd we'd filter out anything else - completed, failed, aborted etc. I would not expect the number of pending/running to be too bad.

Reply:
There is no notification doc for completed states AFAIK.